File: apps/api/src/modules/offers/offer.model.js
import { DataTypes } from 'sequelize';
import sequelize from '../../config/database-mysql.js';

const Offer = sequelize.define('Offer', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
    allowNull: false
  },
  businessId: {
    type: DataTypes.UUID,
    allowNull: true,
    comment: 'FK a business. Null = promo global de plataforma',
    references: {
      model: 'businesses',
      key: 'id'
    },
    onUpdate: 'CASCADE',
    onDelete: 'CASCADE'
  },
  code: {
    type: DataTypes.STRING(50),
    allowNull: false,
    unique: true,
    comment: 'Código único de la promoción'
  },
  title: {
    type: DataTypes.STRING(255),
    allowNull: false
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  discountType: {
    type: DataTypes.ENUM('percentage', 'fixed_amount', 'free_service'),
    allowNull: false,
    comment: 'Tipo de descuento: %, monto fijo, o servicio gratis'
  },
  discountValue: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: false,
    comment: 'Valor del descuento (% o cantidad)'
  },
  maxUses: {
    type: DataTypes.INTEGER,
    allowNull: true,
    comment: 'Máximo de veces que se puede usar. Null = ilimitado'
  },
  usedCount: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
    comment: 'Cuántas veces se ha usado (requiere atomicidad en update)'
  },
  validFrom: {
    type: DataTypes.DATETIME,
    allowNull: false
  },
  validUntil: {
    type: DataTypes.DATETIME,
    allowNull: false
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  },
  applicableServices: {
    type: DataTypes.JSON,
    allowNull: true,
    comment: 'Array de serviceIds a los que aplica. Null = aplica a todos'
  },
  metadata: {
    type: DataTypes.JSON,
    allowNull: true
  }
}, {
  tableName: 'offers',
  timestamps: true,
  indexes: [
    { fields: ['code'], unique: true },
    { fields: ['businessId'] },
    { fields: ['isActive'] },
    { fields: ['validFrom', 'validUntil'] }
  ]
});

export default Offer;

File: apps/mobile/lib/modules/offers/models/offer_model.dart
class OfferModel {
  final String id;
  final String businessId;
  final String code;
  final String description;
  final String discountType; // percentage, fixed, free
  final double discountValue;
  final int? maxUses;
  final int usedCount;
  final DateTime? validFrom;
  final DateTime? validUntil;
  final bool active;
  final DateTime createdAt;

  OfferModel({
    required this.id,
    required this.businessId,
    required this.code,
    required this.description,
    required this.discountType,
    required this.discountValue,
    this.maxUses,
    this.usedCount = 0,
    this.validFrom,
    this.validUntil,
    this.active = true,
    required this.createdAt,
  });

  bool get isValid {
    if (!active) return false;

    final now = DateTime.now();

    if (validFrom != null && now.isBefore(validFrom!)) return false;
    if (validUntil != null && now.isAfter(validUntil!)) return false;

    if (maxUses != null && usedCount >= maxUses!) return false;

    return true;
  }

  factory OfferModel.fromJson(Map<String, dynamic> json) {
    return OfferModel(
      id: json['id'] as String,
      businessId: json['businessId'] as String,
      code: json['code'] as String,
      description: json['description'] as String,
      discountType: json['discountType'] as String,
      discountValue: (json['discountValue'] as num).toDouble(),
      maxUses: json['maxUses'] as int?,
      usedCount: json['usedCount'] as int? ?? 0,
      validFrom: json['validFrom'] is String
          ? DateTime.parse(json['validFrom'] as String)
          : null,
      validUntil: json['validUntil'] is String
          ? DateTime.parse(json['validUntil'] as String)
          : null,
      active: json['active'] as bool? ?? true,
      createdAt: json['createdAt'] is String
          ? DateTime.parse(json['createdAt'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'businessId': businessId,
        'code': code,
        'description': description,
        'discountType': discountType,
        'discountValue': discountValue,
        'maxUses': maxUses,
        'usedCount': usedCount,
        'validFrom': validFrom?.toIso8601String(),
        'validUntil': validUntil?.toIso8601String(),
        'active': active,
        'createdAt': createdAt.toIso8601String(),
      };
}
